subsys/bluetooth/mesh/sensor_cli.c

```diff
@@ -0,0 +1,942 @@
+	  handle_status },
+	{ BT_MESH_SENSOR_OP_COLUMN_STATUS,
+	  BT_MESH_SENSOR_MSG_MINLEN_COLUMN_STATUS, handle_column_status },
+	{ BT_MESH_SENSOR_OP_SERIES_STATUS,
+	  BT_MESH_SENSOR_MSG_MINLEN_SERIES_STATUS, handle_series_status },
+	{ BT_MESH_SENSOR_OP_CADENCE_STATUS,
+	  BT_MESH_SENSOR_MSG_MINLEN_CADENCE_STATUS, handle_cadence_status },
+	{ BT_MESH_SENSOR_OP_SETTINGS_STATUS,
+	  BT_MESH_SENSOR_MSG_MINLEN_SETTINGS_STATUS, handle_settings_status },
+	{ BT_MESH_SENSOR_OP_SETTING_STATUS,
+	  BT_MESH_SENSOR_MSG_MINLEN_SETTING_STATUS, handle_setting_status },
+};
+
+static int sensor_cli_init(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_sensor_cli *cli = mod->user_data;
+
+	cli->mod = mod;
+
+	net_buf_simple_init(cli->pub.msg, 0);
+
+	return 0;
+}
+
+const struct bt_mesh_model_cb _bt_mesh_sensor_cli_cb = {
+	.init = sensor_cli_init,
+};
+
+int bt_mesh_sensor_cli_list_get(struct bt_mesh_sensor_cli *cli,
+				struct bt_mesh_msg_ctx *ctx,
+				struct bt_mesh_sensor_info *sensors,
+				u32_t *count)
+{
+	int err;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_SENSOR_OP_DESCRIPTOR_GET, 0);
+	bt_mesh_model_msg_init(&msg, BT_MESH_SENSOR_OP_DESCRIPTOR_GET);
+
+	struct list_rsp list_rsp = {
+		.sensors = sensors,
+		.count = *count,
+	};
+
+	err = model_ackd_send(cli->mod, ctx, &msg, sensors ? &cli->ack : NULL,
+			      BT_MESH_SENSOR_OP_DESCRIPTOR_STATUS, &list_rsp);
+
+	*count = list_rsp.count;
+	return err;
+}
+
+int bt_mesh_sensor_cli_desc_get(struct bt_mesh_sensor_cli *cli,
+				struct bt_mesh_msg_ctx *ctx,
+				struct bt_mesh_sensor_type *sensor,
+				struct bt_mesh_sensor_descriptor *rsp)
+{
+	int err;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_SENSOR_OP_DESCRIPTOR_GET, 2);
+	bt_mesh_model_msg_init(&msg, BT_MESH_SENSOR_OP_DESCRIPTOR_GET);
+
+	net_buf_simple_add_le16(&msg, sensor->id);
+
+	struct bt_mesh_sensor_info info = {
+		sensor->id,
+	};
+	struct list_rsp list_rsp = {
+		.sensors = &info,
+		.count = 1,
+	};
+
+	err = model_ackd_send(cli->mod, ctx, &msg, rsp ? &cli->ack : NULL,
+			      BT_MESH_SENSOR_OP_DESCRIPTOR_STATUS, &list_rsp);
+	if (err) {
+		return err;
+	}
+
+	if (list_rsp.count == 0) {
+		return -ENOTSUP;
+	}
+
+	if (rsp) {
+		*rsp = info.descriptor;
+	}
+
+	return 0;
+}
+
+int bt_mesh_sensor_cli_cadence_get(struct bt_mesh_sensor_cli *cli,
+				   struct bt_mesh_msg_ctx *ctx,
+				   struct bt_mesh_sensor_type *sensor,
+				   struct bt_mesh_sensor_cadence_status *rsp)
+{
+	if (sensor->channel_count != 1) {
+		return -ENOTSUP;
+	}
+
+	int err;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_SENSOR_OP_CADENCE_GET, 2);
+	bt_mesh_model_msg_init(&msg, BT_MESH_SENSOR_OP_CADENCE_GET);
+
+	net_buf_simple_add_le16(&msg, sensor->id);
+
+	struct cadence_rsp rsp_data = {
+		.id = sensor->id,
+		.cadence = rsp,
+	};
+
+	err = model_ackd_send(cli->mod, ctx, &msg, rsp ? &cli->ack : NULL,
+			       BT_MESH_SENSOR_OP_CADENCE_STATUS, &rsp_data);
+	if (err) {
+		return err;
+	}
+
+	if (!rsp_data.cadence) {
+		return -ENOTSUP;
+	}
+
+	return 0;
```

Comment:
What's going on here?
This is the same as rsp pointer. Which is out variable, setting it to NULL is not supported?

Reply:
Good catch